gnovm/pkg/gnolang/coverage.go

```diff
@@ -0,0 +1,642 @@
+func (c *CoverageData) SaveHTML(outputFileName string) error {
+	tmpl := `
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Coverage Report</title>
+    <style>
+        body { font-family: 'Courier New', Courier, monospace; line-height: 1.5; }
+        .file { margin-bottom: 20px; }
+        .filename { font-weight: bold; margin-bottom: 10px; }
+        pre { margin: 0; }
+        .line { display: flex; }
+        .line-number { color: #999; padding-right: 1em; text-align: right; width: 4em; }
+        .hit-count { color: #666; padding-right: 1em; width: 3em; text-align: right; }
+        .covered { background-color: #90EE90; }
+        .uncovered { background-color: #FFB6C1; }
+    </style>
+</head>
+<body>
+    <h1>Coverage Report</h1>
+    {{range $file, $coverage := .Files}}
+    <div class="file">
+        <div class="filename">{{$file}}</div>
+        <pre>{{range $line, $content := $coverage.Lines}}
+<span class="line{{if $content.Covered}} covered{{else if $content.Executable}} uncovered{{end}}"><span class="line-number">{{$line}}</span><span class="hit-count">{{if $content.Covered}}{{$content.Hits}}{{else}}-{{end}}</span><span class="code">{{$content.Code}}</span></span>{{end}}
+        </pre>
+    </div>
+    {{end}}
+</body>
+</html>`
+
+	t, err := template.New("coverage").Parse(tmpl)
+	if err != nil {
+		return err
+	}
+
+	data := struct {
+		Files map[string]struct {
+			Lines map[int]struct {
+				Code       string
+				Covered    bool
+				Executable bool
+				Hits       int
+			}
+		}
+	}{
+		Files: make(map[string]struct {
+			Lines map[int]struct {
+				Code       string
+				Covered    bool
+				Executable bool
+				Hits       int
+			}
+		}),
+	}
+
+	for path, coverage := range c.Files {
+		realPath, err := c.findAbsoluteFilePath(path)
+		if err != nil {
+			return err
+		}
+		content, err := os.ReadFile(realPath)
+		if err != nil {
+			return err
+		}
+
+		lines := strings.Split(string(content), "\n")
+		fileData := struct {
+			Lines map[int]struct {
+				Code       string
+				Covered    bool
+				Executable bool
+				Hits       int
+			}
+		}{
+			Lines: make(map[int]struct {
+				Code       string
+				Covered    bool
+				Executable bool
+				Hits       int
+			}),
+		}
+
+		for i, line := range lines {
+			lineNum := i + 1
+			hits, covered := coverage.HitLines[lineNum]
+			executable := coverage.ExecutableLines[lineNum]
+
+			fileData.Lines[lineNum] = struct {
+				Code       string
+				Covered    bool
+				Executable bool
+				Hits       int
+			}{
+				Code:       line,
+				Covered:    covered,
+				Executable: executable,
+				Hits:       hits,
+			}
+		}
+
+		data.Files[path] = fileData
+	}
+
+	file, err := os.Create(outputFileName)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	return t.Execute(file, data)
+}
+
+func (m *Machine) addFileToCodeCoverage(file string, totalLines int) {
```

Comment:
The `Machine` doesn't need this method, function can simply take in `CoverageData`